API Coverage report enhancements

jaydeepk opened the pull request:
**What**:

- The API Coverage report now includes: covered, missed and 'not implemented' endpoints.
- Implemented the ability to exclude certain endpoints from the report
- Implemented the ability to define a success criteria for the report based on the overall api coverage and the total number of missed endpoints. Specmatic can fail the build if the success criteria is not met.



**Checklist**:

- [ ] Documentation added to the README.md OR link to PR on https://github.com/specmatic/specmatic-documentation
- [x] Tests
- [ ] Sonar Quality Gate

jaydeepk committed:
Moved APICoverageReport, APICoverageRow, TestReport classes to separate files.
Extracted the logic to generate the APICoverageReport to a separate class: ApiCoverageReportGenerator(for now).
Added tests for ApiCoverageReportGenerator.

jaydeepk committed:
Minor changes in test for partial coverage.

jaydeepk committed:
Added an 'excludedApis' parameter to ApiCoverageReportGenerator to exclude the specified api from the coverage report.

jaydeepk committed:
Implemented a 'coverage' section in specmatic.json.
It will have properties like: threshold, failBuildIfThresholdNotMet, failBuildIfSpecHasMissedAPIs, excludedAPIs.
This is allow users to configure their build to fail if the coverage if below the threshold value, or if any APIs have been found missing in the specification.
Filtered out the excludedAPIs from the missing APIs reported in the coverage report.

jaydeepk committed:
Refactored the generation of APICoverage rows to potentially cater to generating coverage %age

jaydeepk committed:
Consolidated both covered and missing APIs in a single report.

jaydeepk committed:
Fixed bug in calculating total coverage percentage

vikram-rao committed:
Read report configuration from specmatic.json

vikram-rao committed:
Fixed test failure after adding coverage configuration

jaydeepk committed:
Modified ApiCoverageReportGenerator to accept excluded apis as List<String> as opposed to List<API>.

jaydeepk committed:
Added a basic ReportFormatter interface.
Extracted the text report rendering code in APICoverageReport to a separate CoverageReportTextFormatter class, other minor refactoring.

vikram-rao committed:
Refactor specmatic config file parsing and usage of report configuration information in it in junit support code

jaydeepk committed:
Added method "isBasedOnExample()" to the ContractTest interface.
While calculating coverage, we'll now only consider tests which are based on examples.

jaydeepk committed:
Merge remote-tracking branch 'origin/coverage-report-enhancements' into coverage-report-enhancements

jaydeepk committed:
Integrated the ReportConfiguration in SpecmaticJUnitSupport class to generate coverage report.
Changed minThresholdPercentage and maxMissedEndpointsInSpec to Ints.
Renamed TestReport class to ApiCoverageInput as it is more like the input data needed to generate the coverage report.

jaydeepk committed:
Added assertions to validate the failure criteria for API Coverage report.
Minor refactoring.

jaydeepk committed:
Got the coverage report working even if the whole "report" section is not specified in specmatic.json

jaydeepk committed:
Refactoring

jaydeepk committed:
Added a ReportProcessor class to drive the generation of reports based on configuration.
Made types, APICoverage, OpenAPI fields optional.

jaydeepk committed:
Added a flag in the Scenario class to indicate if the scenario has been generated from examples.
Removed the isBasedOnExample method from ContractTest.
Removed the load method from SpecmaticConfig class which had duplicated logic.

jaydeepk committed:
Moved the reading of report configuration from TestCommand to SpecmaticJUnitSupport.
Made the entire 'types' section mandatory in the report configuration.

jaydeepk committed:
Refactoring, moved the open api coverage generation code to the OpenApiCoverageReportInput class.

jaydeepk committed:
Used the length of the word "response" to render spaces when response code is 0.

jaydeepk committed:
Removed default failure criteria for OpenAPI Coverage configuration.
Handle cases where specmatic.json file does not exist or exists with missing configuration sections.

jaydeepk committed:
Explicitly logged error messages when failure criteria is met

jaydeepk committed:
Minor refactoring, added a ReportProcessor interface.

jaydeepk committed:
Minor refactoring.

jaydeepk committed:
Renamed 'failureCriteria' to successCriteria in the OpenAPI coverage section in specmatic.json.
Asserted on whether all the specified success criteria have been met, instead of asserting each criterion individually.

jaydeepk committed:
If an operation is defined in the spec, but not implemented in code, it won't get counted for coverage.

jaydeepk committed:
Marked all tests which have failed whose path/method are not found in the actuator as not implemented.

jaydeepk committed:
Fixed issues in marking tests as not implemented.

jaydeepk committed:
minor changes

jaydeepk committed:
Renamed the 'count' column to '#excercised' in the API Coverage Summary report.

jaydeepk committed:
Added a 'Remarks' column in the API Coverage report to indicate if the endpoint is covered, missing in spec, or not implemented in code.

jaydeepk committed:
Left aligned all the text columns in the API Coverage report.

jaydeepk committed:
Auto-sized the width of the remarks column based on the longest remarks string.

jaydeepk committed:
clean up

jaydeepk committed:
Merge branch 'main' into coverage-report-enhancements

harikrishnan83 commented on application/src/main/kotlin/application/GraphCommand.kt:
```diff
@@ -1,41 +1,60 @@
             logger = Verbose(CompositePrinter())
 
-        val configJson = SpecmaticConfigJson.load()
+        val configJson = loadSpecmaticJsonConfig(null)
```
are we passing the null intentionally?

jaydeepk replied:
Yeah right now the method expects the 'configFileName' parameter. I'll fix it so that we don't have to pass the null explicitly.

harikrishnan83 commented on core/src/main/kotlin/in/specmatic/core/SpecmaticConfig.kt:
```diff
@@ -85,25 +85,78 @@ data class SpecmaticConfigJson(
 
 @Serializable
-data class RespositoryInfo(
+data class RepositoryInfo(
```
Thanks for fixing the typo

harikrishnan83 commented on junit5-support/src/main/kotlin/in/specmatic/test/SpecmaticJUnitSupport.kt:
```diff
@@ -170,12 +46,30 @@ open class SpecmaticJUnitSupport {
+                ReportTypes(apiCoverage = APICoverage(openAPI = APICoverageConfiguration(successCriteria = SuccessCriteria(0, 0, false))))
+            return when (reportConfiguration) {
+                null -> {
+                    logger.log("Report configuration found empty, hence configuring OpenAPI coverage report without any failure criteria ")
```
Should we update this to "API coverage report configuration not found in specmatic.json, proceeding with API coverage report without success criteria"? This will be in line with the config field name "success criteria".

jaydeepk replied:
Fixed.

harikrishnan83 commented on junit5-support/src/main/kotlin/in/specmatic/test/reports/OpenApiCoverageReportProcessor.kt:
```diff
@@ -0,0 +1,64 @@
+        openApiCoverageReportInput.addExcludedAPIs(reportConfiguration.types.apiCoverage.openAPI.excludedEndpoints)
+        val openAPICoverageReport = openApiCoverageReportInput.generate()
+        if (openAPICoverageReport.rows.isEmpty()) {
+            logger.log("The Open API coverage report generated is blank")
```
Any reason this can happen? should we add what is the corrective action?

jaydeepk replied:
Yes, this can happen in two scenarios:
1. All endpoints are added to the 'excludedEndpoints' array in the 'report' section in specmatic.json.
2. The open api spec does not have any paths documented and the 'endpointsApi' property is not set.

I have updated the message to:
```
The Open API coverage report generated is blank.
This can happen if you have included all the endpoints in the 'excludedEndpoints' array in the report section in specmatic.json, or if your open api specification does not have any paths documented.
```

harikrishnan83 commented on junit5-support/src/main/kotlin/in/specmatic/test/reports/coverage/OpenApiCoverageReportInput.kt:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    fun generate(): OpenAPICoverageReport {
+        // If we get a failed test whose path and method are not found in the actuator, it means that the operation exists in the spec, but has not been implemented.
```
Can we delete comment and extract this section into a method with a name that conveys the same? Example: identifyTestThatFailedBecauseOfEndpointsThatWereNotImplemented

jaydeepk replied:
Fixed.

harikrishnan83 commented on junit5-support/src/main/kotlin/in/specmatic/test/reports/coverage/OpenApiCoverageReportInput.kt:
```diff
@@ -0,0 +1,152 @@
+        var allAPITests = addTestResultsForMissingEndpoints(testResults)
+        allAPITests = sortByPathMethodResponseStatus(allAPITests)
+
+        // Creates a structure which looks like this:
```
Same as above, can we delete comment and extract method?

jaydeepk replied:
Fixed.

harikrishnan83 reviewed (changes requested):
Thanks, looks great.

joelrosario commented on application/build.gradle:
```diff
@@ -59,6 +59,8 @@ dependencies {
 
     implementation 'org.apache.ant:ant-junit:1.10.12'
 
+    implementation "org.jetbrains.kotlinx:kotlinx-serialization-json-jvm:1.4.1"
```
Any reason this needs to be here, it should already be available as a transient dependency of core..?

joelrosario replied:
Not sure if this is needed but I'm merging this for now. We can fix this later.

jaydeepk committed:
Addressed review comments:
1. Added default value(null) for the configFileName param in loadSpecmaticJsonConfig(),  so that we don't explicitly need to pass a null.
2. Improved the info message if no report configuration section is found in specmatic.json.
3. Improved the error message displayed if the API Coverage report generated is blank ( provided guidance on scenarios where a blank report is generated)
4.Extracted code in OpenApiCoverageReportInput.generate() to separate methods with descriptive names to eliminate the need for comments.

Fixed bugs:
1. Division by zero error thrown while calculating the overall API Coverage percentage when the report is empty.
2. The 'excludedEndpoints' specified in specmatic.json were not being excluded from the covered apis.

harikrishnan83 reviewed (approved):
Looks good to me

joelrosario reviewed (approved):
Looks good to me.
